src/eko/io/legacy.py

```diff
@@ -0,0 +1,184 @@
+
+        theory = PseudoTheory.from_old(metaold)
+        operator = PseudoOperator.from_old(metaold)
+
+        # get actual grids
+        arrays = load_arrays(innerdir)
+
+    grid = op5to4(metaold["Q2grid"], arrays)
+
+    with EKO.create(dest) as builder:
+        # here I'm plainly ignoring the static analyzer, the types are faking
+        # the actual ones - not sure if I should fix builder interface to
+        # accept also these
+        eko = builder.load_cards(theory, operator).build()  # pylint: disable=E1101
+
+        for mu2, op in grid.items():
+            eko[mu2] = op
+
+        eko.metadata.version = metaold.get("eko_version", "")
+        eko.metadata.data_version = 0
+        eko.metadata.update()
+
+
+@dataclasses.dataclass
+class PseudoTheory(DictLike):
+    """Fake theory, mocking :cls:`eko.io.runcards.TheoryCard`.
+
+    Used to provide a theory for the :cls:`EKO` builder, even when the theory
+    information is not available.
+
+    """
+
+    _void: str
+
+    @classmethod
+    def from_old(cls, old: RawCard):
+        """Load from old metadata."""
+        return cls(_void="")
+
+
+@dataclasses.dataclass
+class PseudoOperator(DictLike):
+    """Fake operator, mocking :cls:`eko.io.runcards.OperatorCard`.
+
+    Used to provide a theory for the :cls:`EKO` builder, even when the operator
+    information is not fully available.
+
+    """
+
+    mu20: float
+    mu2grid: npt.NDArray
+    rotations: Rotations
+    configs: dict
+
+    @classmethod
+    def from_old(cls, old: RawCard):
+        """Load from old metadata."""
+        mu20 = float(old["q2_ref"])
+        mu2grid = np.array(old["Q2grid"])
+
+        xgrid = XGrid(old["interpolation_xgrid"])
+        pids = old.get("pids", np.array(br.flavor_basis_pids))
+
+        rotations = Rotations(xgrid=xgrid, pids=pids)
+
+        def set_if_different(name: str, default: npt.NDArray):
+            basis = old.get(name)
+            if basis is not None and not np.allclose(basis, default):
+                setattr(rotations, name, basis)
+
+        set_if_different("inputpids", pids)
+        set_if_different("targetpids", pids)
+        set_if_different("inputgrid", xgrid.raw)
+        set_if_different("targetgrid", xgrid.raw)
+
+        configs = dict(
+            interpolation_polynomial_degree=old.get("interpolation_polynomial_degree"),
+            interpolation_is_log=old.get("interpolation_is_log"),
+        )
+
+        return cls(mu20=mu20, mu2grid=mu2grid, rotations=rotations, configs=configs)
+
+
+ARRAY_SUFFIX = ".npy.lz4"
+"""Suffix for array files inside the tarball."""
+
+
+def load_arrays(dir: pathlib.Path) -> dict:
+    """Load arrays from compressed dumps."""
+    arrays = {}
+    for fp in dir.glob(f"*{ARRAY_SUFFIX}"):
+        with lz4.frame.open(fp, "rb") as fd:
+            # static analyzer can not guarantee the content to be bytes
+            content = fd.read()
+            if isinstance(content, str):
```

Comment:
It will never occur, since we are writing bytes, but `lz4.frame.open().read()` can return both, according to the string you put in the second argument of `open`. In this case is `"rb"` so it will be forever bytes.

However, a string literal is a value, and PyRight (same for mypy) cannot rely on values (that only exist at runtime) for the static analysis (`"rb"` is hardcoded, but in order to discriminate the behavior according to that value it should execute `open`, so it is only possible at runtime).
Thus is not able to guarantee that it will be bytes, and I'm telling it through `isinstance`.

Essentially, I want to enforce more Type Hints, and make better use of the editor static checks (in practice, it prevents the red underlining in VS code).